src/ngraph/frontend/onnxifi/tensor.hpp

```diff
@@ -0,0 +1,141 @@
+                case ONNXIFI_DATATYPE_INT8:
+                case ONNXIFI_DATATYPE_INT16:
+                case ONNXIFI_DATATYPE_INT32:
+                case ONNXIFI_DATATYPE_INT64:
+                case ONNXIFI_DATATYPE_UINT8:
+                case ONNXIFI_DATATYPE_UINT16:
+                case ONNXIFI_DATATYPE_UINT32:
+                case ONNXIFI_DATATYPE_UINT64: break;
+                case ONNXIFI_DATATYPE_COMPLEX64:
+                case ONNXIFI_DATATYPE_COMPLEX128: throw status::invalid_datatype{};
+                default: throw status::unsupported_datatype{};
+                }
+                switch (tensor.memoryType)
+                {
+                case ONNXIFI_MEMORY_TYPE_CPU: break;
+                case ONNXIFI_MEMORY_TYPE_CUDA_BUFFER:
+                case ONNXIFI_MEMORY_TYPE_OPENCL_BUFFER:
+                case ONNXIFI_MEMORY_TYPE_OPENGLES_TEXTURE_2D:
+                case ONNXIFI_MEMORY_TYPE_D3D_RESOURCE: throw status::invalid_memory_type{};
+                default: throw status::unsupported_memory_type{};
+                }
+                if ((tensor.dimensions != 0) && (tensor.shape == nullptr))
+                {
+                    throw status::null_pointer{};
+                }
+                if ((tensor.shape != nullptr) && (tensor.dimensions == 0))
+                {
+                    throw status::invalid_size{};
+                }
+                if (tensor.shape == nullptr)
+                {
+                    m_shape = {1};
+                }
+                else
+                {
+                    Span<uint64_t> shape{tensor.shape, tensor.dimensions};
+                    for (const auto& value : shape)
+                    {
+                        if (value == 0)
+                        {
+                            throw status::invalid_shape{};
+                        }
+                        m_shape.push_back(value);
+                        m_size *= value;
+                    }
+                }
+                if (tensor.buffer == 0)
+                {
+                    throw status::invalid_memory_location{};
+                }
+            }
+
+            virtual std::shared_ptr<runtime::Tensor> to_ng(runtime::Backend& backend) const = 0;
+
+            const void* data() const { return reinterpret_cast<const void*>(m_tensor->buffer); }
+            std::size_t size() const { return m_size; }
+            const Shape& get_shape() const { return m_shape; }
+            const char* get_name() const { return m_tensor->name; }
+        protected:
+            const ::onnxTensorDescriptorV1* m_tensor;
+            Shape m_shape;
+            std::size_t m_size{1};
+        };
+
+        class InputTensor final : public Tensor
+        {
+        public:
+            using Tensor::Tensor;
+            std::shared_ptr<runtime::Tensor> to_ng(runtime::Backend& backend) const final;
+        };
+
+        class OutputTensor final : public Tensor
```

Comment:
Please split the `tensor` file into separate files, one for each class.

Reply:
Spliting into separate files will add complexity and increase compilation time due to extra preprocessor passes required to include those 5 lines + comple header comment + extra `#pragma one` guard stored in cache... `Tensor`, `OutputTensor` and `InputTensor` are intentionally placed in the single header file.

Reply:
Those 3 classes are logically and conceptually consistent and I prefer to keep it as is.

Reply:
I would still rather see the classes split into separate files for easier navigation when editing code.

Reply:
done